johnzon-core/src/main/java/org/apache/johnzon/core/Snippet.java

```diff
@@ -0,0 +1,435 @@
+
+        private void write(final String name, final JsonValue value) {
+            if (snippet.terminate()) {
+                return;
+            }
+
+            switch (value.getValueType()) {
+                case ARRAY:
+                    generator.writeStartArray(name);
+                    generator.flush();
+                    final JsonArray array = value.asJsonArray();
+                    for (final JsonValue jsonValue : array) {
+                        if (snippet.terminate()) {
+                            break;
+                        }
+                        write(jsonValue);
+                    }
+                    generator.writeEnd();
+                    generator.flush();
+
+                    break;
+                case OBJECT:
+                    generator.writeStartObject(name);
+                    generator.flush();
+                    final JsonObject object = value.asJsonObject();
+                    for (final Map.Entry<String, JsonValue> keyval : object.entrySet()) {
+                        if (snippet.terminate()) {
+                            break;
+                        }
+                        write(keyval.getKey(), keyval.getValue());
+                    }
+                    generator.writeEnd();
+                    generator.flush();
+
+                    break;
+                default: {
+                    generator.write(name, value);
+                    generator.flush();
+                }
+            }
+        }
+
+        private String get() {
+            generator.flush();
+            return snippet.isTruncated() ? snippet.get() + "..." : snippet.get();
+        }
+
+        @Override
+        public void close() {
+            generator.close();
+        }
+    }
+
+    /**
+     * Specialized OutputStream with three internal states:
+     * Writing, Completed, Truncated.
+     *
+     * When there is still space left for more json, the
+     * state will be Writing
+     *
+     * If the last write brought is exactly to the end of
+     * the max length, the state will be Completed.
+     *
+     * If the last write brought us over the max length, the
+     * state will be Truncated.
+     */
+    static class SnippetOutputStream extends OutputStream {
+
+        private final ByteArrayOutputStream buffer;
+        private OutputStream mode;
+
+        public SnippetOutputStream(final int max) {
+            this.buffer = new ByteArrayOutputStream(Math.min(max, 8192));
+            this.mode = new Writing(max, buffer);
+        }
+
+        public String get() {
+            return buffer.toString();
+        }
+
+        /**
+         * Calling this method implies the need to continue
+         * writing and a question on if that is ok.
+         *
+         * It impacts internal state in the same way as
+         * calling a write method.
+         *
+         * @return true if no more writes are possible
+         */
+        public boolean terminate() {
```

Comment:
Like this rework, way more explicit!

Reply:
Took a lot of renaming till I found something just right :)